src/structure/pdb.jl

```diff
@@ -30,26 +49,315 @@ end
+end
+
+
+"""
+Returns a list of all obsolete entries ever in the RCSB PDB server.
+"""
+function pdbobsoletelist()
+    obsoletelist = String[]
+    info("Fetching list of all obsolete PDB Entries from RCSB PDB Server...")
+    tempfilepath = tempname()
+    download("ftp://ftp.wwpdb.org/pub/pdb/data/status/obsolete.dat", tempfilepath)
+    open(tempfilepath) do input
+        for line in eachline(input)
+            # Check if its an obsolete pdb entry and not headers
+            if line[1:6] == "OBSLTE"
+                # The 21st to 24th characters in obsolete pdb entry has the pdb id
+                pdbid = uppercase(line[21:24])
+                # Check PDB ID is 4 characters long and only consits of alphanumeric characters
+                if length(pdbid) != 4 || ismatch(r"[^a-zA-Z0-9]", pdbid)
+                    throw(ArgumentError("Not a valid PDB ID: \"$pdbid\""))
+                end
+                push!(obsoletelist,pdbid)
+            end
+        end
+    end
+    rm(tempfilepath)
+    return obsoletelist
+end
+
+
+"""
+Download a PDB file or biological assembly from the RCSB PDB server. 
+By default downloads the PDB file to current working directory; 
+if the keyword argument `ba_number` is set the biological assembly with that
+number will be downloaded; 
+if the keyword argument `pdb_dir` is set the PDB file is downloaded to the
+specified directory; 
+if the keyword argument `obsolete` is set `true`, the PDB file is downloaded
+into the obsolete directory inside `pdb_dir`;
+if the keyword argument `overwrite` is set `true`, then it will overwrite the
+PDB file if it exists in the `pdb_dir`.
+"""
+function downloadpdb(pdbid::AbstractString; pdb_dir::AbstractString=pwd(), file_format::Type=PDB, obsolete::Bool=false, overwrite::Bool=false, ba_number::Integer=0)
     # Check PDB ID is 4 characters long and only consits of alphanumeric characters
     if length(pdbid) != 4 || ismatch(r"[^a-zA-Z0-9]", pdbid)
         throw(ArgumentError("Not a valid PDB ID: \"$pdbid\""))
+    end 
+    # Check if the PDB file is marked as obsolete
+    if obsolete
+        # Set the download path to obsolete directory inside the "pdb_dir"
+        pdb_dir = joinpath(pdb_dir,"obsolete")
+    end
+    # Check and create directory if it does not exists in filesystem
+    if !isdir(pdb_dir) 
+        info("Creating directory : $pdb_dir")
+        mkpath(pdb_dir)
     end
-    if ba_number == 0
-        download("http://www.rcsb.org/pdb/files/$pdbid.pdb", out_filepath)
+    archivefilepath = tempname()
+    # The PDB file name will end with the ba_number when downloading biological assembly
+    if ba_number==0
+        pdbpath = joinpath(pdb_dir,"$pdbid"*pdbextension[file_format])
     else
-        # Will download error page if ba_number is too high
-        download("http://www.rcsb.org/pdb/files/$pdbid.pdb$ba_number", out_filepath)
+        pdbpath = joinpath(pdb_dir,"$pdbid"*pdbextension[file_format]*"$ba_number")
```

Comment:
Thinking a bit more about it, the biological number should not change the extension of the file. Ideally it would change the filename, e.g. `1ABC_ba1.pdb` or something like that rather than `1ABC.pdb1`.